### PR TITLE
Adding Winograd Coreference

### PR DESCRIPTION
**Pull request by pruksmhc**
Added components: 
-A preprocessing script for winograd with tests
-Added superglue to the list in tasks/__init__.py 
-Adding to tasks.py

### CODE REVIEW

File: scripts/winograd/preprocess_winograd.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env python
+
+"""
+This file will preprocess Winograd, aligning the span indices to the tokenizaer of
```
Comment (sleepinyourhat) [inline scripts/winograd/preprocess_winograd.py]: 'preprocess the SuperGLUE Winograd Schema Challenge data,'
There's no standard format for Winograd schemas, so need to specify.

File: src/tasks/tasks.py [hunk 2]
```diff
@@ -2224,3 +2224,26 @@ def get_metrics(self, reset=False):
...
+class WinogradCoreferenceTask(SpanClassificationTask):
+    def __init__(self, path,  **kw):
+        self._files_by_split = {'train': "train.json", 'val': "val.json",'test': "test_final.json"}
+        self.num_spans = 2
+        super().__init__(files_by_split=self._files_by_split, label_file="labels.txt", path=path, **kw)
+        self.val_metric = "%s_acc" % self.name 
+
+    def update_metrics(self, logits, labels, tagmask=None):
+        logits, labels = logits.detach(), labels.detach()
+        def one_hot_v(batch, depth=2):
```
Comment (sleepinyourhat) [inline src/tasks/tasks.py]: Comment: What does this do?

File: src/utils/utils.py [hunk 3]
```diff
@@ -188,7 +188,7 @@ def split_data(data, ratio, shuffle=1):
         splits[1].append(col[split_pt:])
     return tuple(splits[0]), tuple(splits[1])
     
-def unbind_predictions(self, preds: torch.Tensor) -> Iterable[np.ndarray]:
+def unbind_predictions(preds: torch.Tensor) -> Iterable[np.ndarray]:
```
Comment (sleepinyourhat) [inline src/utils/utils.py]: What's going on here? This method does get called in a few places, so this change should break the repo...

File: src/tasks/tasks.py [hunk 4]
```diff
@@ -2224,3 +2224,26 @@ def get_metrics(self, reset=False):
         f1 = (f11 + f12 + f13) / 3
         return {'accuracy': acc, 'f1': f1, 'precision': pcs, 'recall': rcl}
 
+@register_task('winograd-coreference', rel_path = 'winograd-coref')
+class WinogradCoreferenceTask(SpanClassificationTask):
+    def __init__(self, path,  **kw):
+        self._files_by_split = {'train': "train.json", 'val': "val.json",'test': "test_final.json"}
```
Comment (W4ngatang) [inline src/tasks/tasks.py]: these should be jsonl

File: src/utils/utils.py (hunk 3 above)
Comment (W4ngatang) [reply src/utils/utils.py]: It was previously a task method (that didn't need to be) and is pulled out, so this deletes `self`, but idk why it wasn't raising issues earlier.

File: src/tasks/tasks.py (hunk 4 above)
Comment (pruksmhc) [reply src/tasks/tasks.py]: After running the preprocessing script, you will get JSON data instead of jsonl. I can change this, but don't think it's too urgent.

File: src/tasks/tasks.py (hunk 2 above)
Comment (pruksmhc) [reply src/tasks/tasks.py]: This creates one hot embeddings given the logits , which is what the F1 scorer and acc scorer expects.

Comment (sleepinyourhat) [reply src/tasks/tasks.py]: Just suggesting you comment in the code.

File: src/utils/utils.py (hunk 3 above)
Comment (sleepinyourhat) [reply src/utils/utils.py]: But isn't it called with [something].unbind_predictions() elsewhere? Won't that break?

Comment (sleepinyourhat) [reply src/utils/utils.py]: @pruksmhc Can you confirm that all the calls to this are non-broken?

File: src/tasks/__init__.py [hunk 5]
```diff
@@ -26,6 +26,8 @@
 
 ALL_SUPERGLUE_TASKS = ["commitbank", "wic"]
 
+ALL_SUPERGLUE_TASKS = ["winograd-coreference"]
```
Comment (sleepinyourhat) [inline src/tasks/__init__.py]: !!

File: src/tasks/tasks.py [hunk 6]
```diff
@@ -2651,7 +2651,41 @@ def _make_instance(question, choices, label, idx):
...
+            ones = torch.sparse.torch.eye(depth).cuda()
+            return ones.index_select(0, batch)
+
+        binary_preds = make_one_hot(logits, depth=2)
+        # Make label_ints a batch_size list of labels
+        label_ints = torch.argmax(labels, dim=1)
+        self.f1_scorer(binary_preds, label_ints)
+        self.acc_scorer(binary_preds.long(), labels.long())
+
     def get_metrics(self, reset=False):
```
Comment (sleepinyourhat) [inline src/tasks/tasks.py]: It looks like get_metrics for the previous task got cut off in a merge fix.

File: src/utils/utils.py (hunk 3 above)
Comment (sleepinyourhat) [reply src/utils/utils.py]: ...confirmed?
